packages/multi-tenant/src/discoverTenant.ts

```diff
@@ -0,0 +1,38 @@
+import getMatch from "./lib/getMatch";
+import getMultiTenantConfig from "./lib/multiTenantConfig";
+import TenantService from "./model/tenants/service";
+
+import type { Tenant } from "./types";
+import type { FastifyReply, FastifyRequest } from "fastify";
+
+const discoverTenant = async (request: FastifyRequest, reply: FastifyReply) => {
+  const { config, slonik, sql } = request;
+
+  const { slugs: reservedSlugs, domains: reservedDomains } =
+    getMultiTenantConfig(config).reserved;
+
+  const { matchedDomain, matchedSlug } = getMatch(request);
+
+  if (
+    !reservedDomains.includes(matchedDomain) ||
+    !reservedSlugs.includes(matchedSlug)
+  ) {
+    if (!matchedSlug) {
+      reply.send({
```

Comment:
This function should not be responsible for sending a response. That's the job of the plugin. Instead this method should return a tenant, or null, or throw an error.
You need to analyze the possible outcomes, and determine the appropriate output:
* if reserved domain, return null
* else if no tenant found, throw error
* else return tenant